include/nuttx/semaphore.h

```diff
@@ -45,23 +45,42 @@
 /* semcount, flags, waitlist, hhead */
 
 #    define NXSEM_INITIALIZER(c, f) \
-       {(c), (f), SEM_WAITLIST_INITIALIZER, NULL}
+       {{c}, (f), SEM_WAITLIST_INITIALIZER, NULL}
 #  else
 /* semcount, flags, waitlist, holder[2] */
 
 #    define NXSEM_INITIALIZER(c, f) \
-       {(c), (f), SEM_WAITLIST_INITIALIZER, SEMHOLDER_INITIALIZER}
+       {{(c)}, (f), SEM_WAITLIST_INITIALIZER, SEMHOLDER_INITIALIZER}
 #  endif
 #else /* CONFIG_PRIORITY_INHERITANCE */
 /* semcount, flags, waitlist */
 
 #  define NXSEM_INITIALIZER(c, f) \
-     {(c), (f), SEM_WAITLIST_INITIALIZER}
+     {{(c)}, (f), SEM_WAITLIST_INITIALIZER}
 #endif /* CONFIG_PRIORITY_INHERITANCE */
 
-/* Macro to retrieve sem count */
+/* Macros to retrieve sem count and to check if nxsem is mutex */
 
-#define NXSEM_COUNT(s) ((FAR atomic_t *)&(s)->semcount)
+#define NXSEM_COUNT(s)        ((FAR atomic_t *)&(s)->val.semcount)
+#define NXSEM_IS_MUTEX(s)     (((s)->flags & SEM_TYPE_MUTEX) != 0)
+
+/* Mutex related helper macros */
+
+#define NXSEM_MBLOCKS_BIT    (((uint32_t)1) << 31)
+#define NXSEM_NO_MHOLDER     ((uint32_t)0x7ffffffe)
```

Comment:
Is there some reason 0 cannot be used here ? A process can never get pid = 0.

Reply:
It happens in the early boot afaik

Reply:
Yes the OS init phase came to mind but I'm fairly certain it should not happen. There should be logic that prevents calls to nxsem/nxmutex_wait from happening.

Reply:
There are still calls happening in OS init phase, I don't now remember where exactly. Also, there were other reasons not to change this, see below. -1 is just as good number as some other, and in this case it is even better as it is the same as before

Reply:
-1 is fine, I was just wondering if it could be 0 since that feels more natural for me for some reason :P

Reply:
quick snippet from pthread_mutexconsistent.c:
```
     pid = mutex_get_holder(&mutex->mutex);

      /* Is the mutex available? */

      DEBUGASSERT(pid != 0); /* < 0: available, >0 owned, ==0 error */
      if (pid >= 0)
```

Similar places around...